test/jdk/java/net/httpclient/AuthFilter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.Authenticator;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+
+/**
+ * @test
+ * @bug 8263442
+ * @summary Potential bug in jdk.internal.net.http.common.Utils.CONTEXT_RESTRICTED
+ * @run main/othervm AuthFilter
+ */
+
+public class AuthFilter {
+    static class Auth extends Authenticator {
+    }
+
+    static HttpServer createServer() throws IOException {
+        HttpServer server = HttpServer.create(new InetSocketAddress(0), 5);
+        HttpHandler handler = (HttpExchange e) -> {
+            InputStream is = e.getRequestBody();
+            is.readAllBytes();
+            is.close();
+            Headers reqh = e.getRequestHeaders();
+            if (reqh.containsKey("authorization")) {
+                e.sendResponseHeaders(500, -1);
```

Comment:
I am a bit concerned by that. It shows that without your fix preemptive authentication would have worked, as the server would have received the authentication header.

I did a bit of an experiment - and it seems that with proxy-authorization you would get an IOException (with or without your fix). So it seems that without your fix we are unwillingly currently supporting user preemptive  authentication (for servers) but not for proxies. With your fix, neither will be supported.

Is that the right thing to do?

Reply:
What I am seeing is that if no authenticator set, whether the fix is present or not, an "Authorization" header is passed through, but a "Proxy-Authorization" header is filtered. So, that is a different issue. It probably is a bug though.

Reply:
I've updated the test to test the proxy authorization case

Reply:
Don't you need to set `-Djdk.http.auth.proxying.disabledSchemes=""` on the command line to have a chance that Proxy-Authorization will be forwarded?